// @jakeyprime/logger/index.js
const chalk = require('chalk');
const fs = require('fs');
const { inspect } = require('util');

class Logger {
  constructor() {
    this.level = process.env.LEVEL || process.env.level || 'info';
    this.levels = {
      debug: 0,
      info: 1,
      warn: 2,
      error: 3,
    };
  }

  time() {
    const date = new Date();
    const hours =
      `${date.getHours()}`.length === 1
        ? `0${date.getHours()}`
        : date.getHours();
    const minutes =
      `${date.getMinutes()}`.length === 1
        ? `0${date.getMinutes()}`
        : date.getMinutes();
    const seconds =
      `${date.getSeconds()}`.length === 1
        ? `0${date.getSeconds()}`
        : date.getSeconds();
    return `${hours}:${minutes}:${seconds}`;
  }

  write(type, message) {
    const date = new Date();
    fs.appendFileSync(
      'events.log',
      `\n${date.getFullYear()}/${date.getMonth()}/${date.getDate()} ${this.time()} [${type}] ${message}`,
    );
  }

  setLevel(level) {
    this.level = level;
    return this;
  }

  debug(message) {
    if (this.levels[this.level] > 1) return;
    if (typeof message === 'object') message = inspect(message);
    const type = ' debug ';
    this.write(type, message);
    console.log(`${this.time()} [${chalk.gray(type)}] ${chalk.gray(message)}`);
  }

  info(message) {
    if (this.levels[this.level] > 1) return;
    if (typeof message === 'object') message = inspect(message);
    const type = ' info  ';
    this.write(type, message);
    console.log(`${this.time()} [${chalk.cyan(type)}] ${message}`);
  }

  warn(message) {
    if (this.levels[this.level] > 2) return;
    if (typeof message === 'object') message = inspect(message);
    const type = 'warning';
    this.write(type, message);
    console.log(`${this.time()} [${chalk.yellow(type)}] ${message}`);
  }

  error(message) {
    if (this.levels[this.level] > 3) return;
    if (typeof message === 'object') message = inspect(message);
    const type = ' error ';
    this.write(type, message);
    console.error(`${this.time()} [${chalk.red(type)}] ${message}`);
  }
}

module.exports = new Logger();
